File: java-workshop/part-1-powermock/src/main/java/service/DiscountCodeService.java
package service;

import domain.DiscountCode;
import infrastructure.DatabaseDiscountCodeRepository;
import repository.DiscountCodeRepository;

/**
 * Discount code validation service.
 * Instantiates its own repository (consistent with Part 1 architecture).
 */
public class DiscountCodeService {

    private final DiscountCodeRepository repository;

    public DiscountCodeService() {
        // Instantiates repository directly - requires real database connection
        this.repository = new DatabaseDiscountCodeRepository();
    }

    /**
     * Checks a discount code for a given customer.
     *
     * @return true if the code has already been used by this customer
     */
    public boolean checkDiscountCode(String customerId, DiscountCode discountCode) {
        return repository.checkDiscountCode(customerId, discountCode);
    }

    public void markAsUsed(String customerId, DiscountCode discountCode) {
        repository.markAsUsed(customerId, discountCode);
    }
}

File: csharp-workshop/Part1.StaticDeps/Service/DiscountCodeService.cs
using Part1.StaticDeps.Domain;
using Part1.StaticDeps.Infrastructure;
using Part1.StaticDeps.Repository;

namespace Part1.StaticDeps.Service;

/// <summary>
/// Discount code validation service.
/// Instantiates its own repository directly — consistent with Part 1 architecture.
/// </summary>
public class DiscountCodeService
{
    private readonly IDiscountCodeRepository _repository;

    public DiscountCodeService()
    {
        // Instantiates repository directly - requires real database connection
        _repository = new DatabaseDiscountCodeRepository();
    }

    /// <summary>
    /// Checks a discount code for a given customer.
    /// </summary>
    /// <returns>true if the code has already been used by this customer</returns>
    public bool CheckDiscountCode(string customerId, DiscountCode discountCode) =>
        _repository.CheckDiscountCode(customerId, discountCode);

    public void MarkAsUsed(string customerId, DiscountCode discountCode) =>
        _repository.MarkAsUsed(customerId, discountCode);
}
